devito/types/multistage.py

```diff
@@ -0,0 +1,179 @@
+
+
+class MultiStage(Eq):
+    """
+       Abstract base class for multi-stage time integration methods
+       (e.g., Runge-Kutta schemes) in Devito.
+
+       This class represents a symbolic equation of the form `target = rhs`
+       and provides a mechanism to associate it with a time integration
+       scheme. The specific integration behavior must be implemented by
+       subclasses via the `_evaluate` method.
+
+       Parameters
+       ----------
+       lhs : expr-like
+           The left-hand side of the equation, typically a time-updated Function
+           (e.g., `u.forward`).
+       rhs : expr-like, optional
+           The right-hand side of the equation to integrate. Defaults to 0.
+       subdomain : SubDomain, optional
+           A subdomain over which the equation applies.
+       coefficients : dict, optional
+           Optional dictionary of symbolic coefficients for the integration.
+       implicit_dims : tuple, optional
+           Additional dimensions that should be treated implicitly in the equation.
+       **kwargs : dict
+           Additional keyword arguments, such as time integration method selection.
+
+       Notes
+       -----
+       Subclasses must override the `_evaluate()` method to return a sequence
+       of update expressions for each stage in the integration process.
+       """
+
+    def __new__(cls, lhs, rhs=0, subdomain=None, coefficients=None, implicit_dims=None, **kwargs):
+        return super().__new__(cls, lhs, rhs=rhs, subdomain=subdomain, coefficients=coefficients, implicit_dims=implicit_dims, **kwargs)
+
+    def _evaluate(self, **kwargs):
+        raise NotImplementedError(
+            f"_evaluate() must be implemented in the subclass {self.__class__.__name__}")
+
+
+class RK(MultiStage):
+    """
+    Base class for explicit Runge-Kutta (RK) time integration methods defined
+    via a Butcher tableau.
+
+    This class handles the general structure of RK schemes by using
+    the Butcher coefficients (`a`, `b`, `c`) to expand a single equation into
+    a series of intermediate stages followed by a final update. Subclasses
+    must define `a`, `b`, and `c` as class attributes.
+
+    Parameters
+    ----------
+    a : list of list of float
+        The coefficient matrix representing stage dependencies.
+    b : list of float
+        The weights for the final combination step.
+    c : list of float
+        The time shifts for each intermediate stage (often the row sums of `a`).
+
+    Attributes
+    ----------
+    a : list[list[float]]
+        Butcher tableau `a` coefficients (stage coupling).
+    b : list[float]
+        Butcher tableau `b` coefficients (weights for combining stages).
+    c : list[float]
+        Butcher tableau `c` coefficients (stage time positions).
+    s : int
+        Number of stages in the RK method, inferred from `b`.
+    """
+
+    def __init__(self, a=None, b=None, c=None, **kwargs):
```

Comment:
This can just be `def __init__(self, a: list[list[float | np.number]], b: list[float | np.number], c: list[float | np.number], **kwargs) -> None:`, avoiding the need for the validation function